### PR TITLE
[RISCV] Add Smrnmi extension

### PR DESCRIPTION
**Pull request by zhendongmiao**
This commit has completed the Extension for Resumable Non Maskable Interrupts, adding four CRSs and one Trap-Return instruction.
Specification link:["Smrnmi" Extension](https://github.com/riscv/riscv-isa-manual/blob/main/src/rnmi.adoc)

### CODE REVIEW

**Comment by asb** [comment]
Welcome to the project @dong-miao! I'm not sure why the auto-labeler didn't kick in to attach the backend:RISC-V label (this is used to help ping the right people), but I've added that now.

Just two quick comments as it's end of day here so I can't review properly right now:
* We try to make the PR title and description match the intended commit. For a something like this, we'd typically name it something like "[RISCV][MC] Add support for the Smrnmi extension"
* Please add a release note to the RISC-V section in llvm/docs/ReleaseNotes.md. e.g. "Support was added for the Smrnmi extension."

File: llvm/lib/Target/RISCV/RISCVInstrInfo.td [hunk 1]
```diff
@@ -66,6 +66,8 @@ def riscv_sret_glue : SDNode<"RISCVISD::SRET_GLUE", SDTNone,
                              [SDNPHasChain, SDNPOptInGlue]>;
 def riscv_mret_glue : SDNode<"RISCVISD::MRET_GLUE", SDTNone,
                              [SDNPHasChain, SDNPOptInGlue]>;
+def riscv_mnret_glue : SDNode<"RISCVISD::MNRET_GLUE", SDTNone,
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.td:69]: This should be in another patch for CodeGen. And, do we really need it? It will be used in supervisor IIUC.

File: llvm/lib/Target/RISCV/RISCVSystemOperands.td [hunk 2]
```diff
@@ -276,6 +276,14 @@ foreach i = 0...15 in {
 foreach i = 0...63 in
   def : SysReg<"pmpaddr"#i, !add(0x3B0, i)>;
 
+//===----------------------------------------------------------------------===//
+// Machine Non-Maskable Interrupt Handling
+//===----------------------------------------------------------------------===//
+def:SysReg<"mnscratch",0x740>;
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVSystemOperands.td:282]: Add spaces around `:`.

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVSystemOperands.td:282]: ARen't these already in the file?

```
//===-----------------------------------------------                             
// Resumable Non-Maskable Interrupts(Smrnmi) CSRs                                
//===-----------------------------------------------                             
def : SysReg<"mnscratch", 0x740>;                                                
def : SysReg<"mnepc", 0x741>;                                                    
def : SysReg<"mncause", 0x742>;                                                  
def : SysReg<"mnstatus", 0x744>;
```

File: llvm/test/MC/RISCV/machine-csr-names.s [hunk 3]
```diff
@@ -1913,6 +1913,66 @@ csrrs t1, mhpmcounter31, zero
 csrrs t2, 0xB1F, zero
 
 
+######################################
+# Machine Counter Setup
```
Comment (topperc) [inline llvm/test/MC/RISCV/machine-csr-names.s:1917]: Section title is incorrect.

The registers are already tested in rv32-machine-csr-names.s but they should be in this file. That was a mistake.

File: llvm/lib/Target/RISCV/RISCVSystemOperands.td (hunk 2 above)
Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVSystemOperands.td:282]: Yes! I forgot it.

File: llvm/lib/Target/RISCV/RISCVInstrInfo.td [hunk 4]
```diff
@@ -813,6 +815,12 @@ def MRET : Priv<"mret", 0b0011000>, Sched<[]> {
   let rs1 = 0;
   let rs2 = 0b00010;
 }
+
+def MNRET : Priv<"mnret", 0b0111000>, Sched<[]> {
```
Comment (lenary) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.td:819]: Does this need a `Requires=[HasStdExtSmrnmi]` (And a definition of `HasStdExtSmrnmi`)?

File: llvm/lib/Target/RISCV/RISCVFeatures.td [hunk 5]
```diff
@@ -931,6 +931,10 @@ def FeatureStdExtSmepmp
     : RISCVExtension<"smepmp", 1, 0,
                      "'Smepmp' (Enhanced Physical Memory Protection)">;
 
+def FeatureStdExtSmrnmi
+    : RISCVExtension<"smrnmi", 1, 0,
+                     "'Smrnmi' (Extension for Resumable Non-Maskable Interrupts)">;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFeatures.td:936]: I don't think we usually write "Extension for" in the description.

File: llvm/lib/Target/RISCV/RISCVInstrInfo.td (hunk 1 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.td:69]: RISCVISD::MNRET_GLUE isn't defined in this patch so it should definitely be removed.

File: llvm/lib/Target/RISCV/RISCVFeatures.td (hunk 5 above)
Comment (dong-miao) [reply llvm/lib/Target/RISCV/RISCVFeatures.td]: Okay, I have deleted the relevant content.

File: llvm/lib/Target/RISCV/RISCVInstrInfo.td (hunk 4 above)
Comment (dong-miao) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.td:819]: Thanks for your help. I have added the relevant content.

File: llvm/lib/Target/RISCV/RISCVInstrInfo.td (hunk 1 above)
Comment (dong-miao) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.td]: Thanks, I have removed the relevant content.

File: llvm/docs/RISCVUsage.rst [hunk 6]
```diff
@@ -129,6 +129,7 @@ on support follow.
      ``Smcdeleg``      Supported
      ``Smcsrind``      Supported
      ``Smepmp``        Supported
+     ``Smrnmi``        Supported
```
Comment (lenary) [inline llvm/docs/RISCVUsage.rst:132]: This should probably be "Assembly Support" - some of the other `*ret` instructions have support in clang for `__attribute((interrupt(<kind>)))` which doesn't yet exist for `mnret` - there's no obligation to add this support in the first PR, but this should mean no one expects it has the same level of featureas as the other `*ret` instructions.
```suggestion
     ``Smrnmi``        Assembly Support
```

File: llvm/lib/Target/RISCV/RISCVInstrInfo.td (hunk 4 above)
Comment (lenary) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.td:819]: Nice, thanks!

File: llvm/docs/RISCVUsage.rst (hunk 6 above)
Comment (dong-miao) [reply llvm/docs/RISCVUsage.rst]: Okay, thanks for your help.

**Review by lenary — APPROVED** [review]
LGTM.

**Comment by Xinlong-Wu** [comment]
@dong-miao don‘t have write access, I help him to merge
